src/composer.ts

```diff
@@ -34,6 +42,38 @@ function getText(
   return undefined
 }
 
+type Props = {
+  [key in tt.UpdateType]: tt.UpdateProps[key] & tt.ContextProps[key]
+}
+type SubProps = {
+  [key in tt.MessageSubType]: Props['message'] &
+    tt.UpdateSubProps[key] &
+    tt.ContextSubProps[key]
+}
+
+type MatchedContext<
+  C extends Context,
+  T extends tt.UpdateType | tt.MessageSubType
+> = C &
+  // TODO: support subtypes for: callback query, channel post, maybe more?
```

Comment:
Do we want to support type inference in even more cases, not just for subtypes of `message`?

Reply:
Currently, when using a `MessageSubType`, we infer subtypes for the `message` property only, even though it could come from a different update type.

Take this code:
```ts
bot.on('text', ctx => {
  const text = ctx.message.text // works
})
```
We correctly infer `ctx.message.text: string` as well as `ctx.update.message.text: string`.

However, `mount` not only checks message texts, but also the content of all update types, including callback queries, inline queries, channel posts, etc.

Take this code:
bot.on('channel_chat_created', ctx => {
  const chat = ctx.channelPost.chat // error
})
We infer both `ctx.message.channel_chat_created` and `ctx.update.message.channel_chat_created` as `true` and `ctx.channelPost` as `undefined`, all of which is incorrect.

I'm suggesting that we should also do a similar thing for other subtypes that we provide. It may make sense to do this generically, but I believe that this will create too complex types again, so we may have to come up with something smarter.

Note: the TODO comment in the code is not quite correct about callback queries as neither `'data'` nor `'game_short_name`' are valid message subtypes, so callback queries should in fact not be the issue here. It's about channel posts, edits, and channel edits.

Reply:
Now that I think about it, it basically boils down to supporting channel mode also in this place.